### ISSUE TITLE
Feature Request - agent attributes

### ISSUE THREAD

**Issue by dustymc** [ArctosDB/arctos#7318]
**Is your feature request related to a problem? Please describe.**

https://github.com/ArctosDB/arctos/issues/6813 - agent goals have clarified and a new thing is needed to meet them
https://github.com/ArctosDB/arctos/issues/5172 - a need for more link between various agent-things has emerged

**Describe what you're trying to accomplish** 

Ultimately unify agent data into one structure capable of carrying and versioning more, and more related (or minimal when necessary) agent information. For now, preview the structure with shinies.

**Describe the solution you'd like**

Something like https://github.com/ArctosDB/arctos/issues/5172 (initially accommodating only https://github.com/ArctosDB/arctos/issues/6813#issuecomment-1899522184)

**Describe alternatives you've considered**

Don't think there are any which align with the goals.

**Additional context**

I believe this should be viewed as entirely additive and so needing minimal further discussion at this point. However, the next phase will involve migrating relationships, statuses, and addresses to this unified structure (where they can be better integrated, and changes can be monitored). Can we somehow communicate that to The Community? I can't tell if everyone is paying attention and likes where this is going or ????

**Priority**

Critical?

@Jegelewicz @mkoo plz confirm I'm not lost and I'll get started.

**Comment by Jegelewicz** [comment]
That all makes sense to me - we should discuss at the AWG meeting on Thursday.

**Comment by AJLinn** [comment]
I'd appreciate a mockup/demonstration about the proposed updates at the AWG meeting. This all happened very quickly amidst about a thousand code table updates (LOL, those damned lithostratigraphic  adds!) so I admit I haven't been able to follow all the new & linked issues and discussions that followed. Thanks in advance!

**Comment by dustymc** [comment]
@AJLinn is there something missing from @Jegelewicz 's[ mockup/spreadsheet](https://docs.google.com/spreadsheets/d/1h0cxNvMyIJuM9M30g75t5hG7k6V1FbxEOlUb7c7cnvA/edit#gid=0), or what can we do to clarify?

>lithostratigraphic

I filter email notifications, Teresa can probably demo filtering in github.  You could also bookmark something like https://github.com/ArctosDB/arctos/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+-label%3Astratigraphy 

>all the new & linked issues

All Agent issues should be wearing a label, which makes them available as https://github.com/ArctosDB/arctos/labels/Function-Agents

I've also tried to summarize my understanding at the top of https://github.com/ArctosDB/arctos/issues/6813, and in https://docs.google.com/document/d/1tSk344mJsIRYZ-BJcYfyaPEmbrFS-wSqZzJ5dO1E2EU/edit#heading=h.ei43n7pwmxwu

**Comment by dustymc** [comment]
AWG - short discussion - GO!

**Comment by dustymc** [comment]
This could use feedback. Do please note that when this goes to production we will again be "in the betweens" - this is running in addition to normal/legacy agent data and structure, everything is going to be a bit clunky as long as that's the case, the goal should be to find any problems with the new structure/tools and move forward with getting everything into it ASAP.

As above, this is just tacked on to the bottom of the normal edit page - the "normal" buttons do nothing for attributes, the 'save' button saves only attributes and ignores any changes above.

The form is set up so that three new attributes may be added at a time, that's easy to adjust to any number. 

The layout is responsive, it should work reasonably well on most any device, and it's designed such that it could be improved for any class of device.

I have split status out as a demonstration of the flexibility of the model and UI; the data are stored with everything else. Status also has some special database-level rules. (We may want to expand those to other attributes at some point, but at the moment I'm thinking that a maximally-flexible/minimally-constrained model is probably most useful.)

There's some styling based on status. 

Nobody has update or delete on the table, and there's a UI dedicated to viewing 'deprecated' data.

Updates are running through an API; if you don't like my UI you can write your own.

What could be better?

**Comment by ewommack** [comment]
> What could be better?

Ok so test by:
1. creating agents
2. searching for agents
3. using agents in Arctos tools

Try and break or just try and do everything with?

**Comment by dustymc** [comment]
>test by:

None of that is involved in this. It's just a bit of new UI hanging precariously off the bottom of the edit agent page. I just need to know if that's going to work at this point, then I'll start filing issues to move all the normal agent-stuff into it (and then that stuff will need tested).

**Comment by ewommack** [comment]
>  I just need to know if that's going to work at this point

So what can we do to let you know if it is working?

**Comment by AJLinn** [comment]
> @AJLinn is there something missing from @Jegelewicz 's[ mockup/spreadsheet](https://docs.google.com/spreadsheets/d/1h0cxNvMyIJuM9M30g75t5hG7k6V1FbxEOlUb7c7cnvA/edit#gid=0), or what can we do to clarify?

Sorry for the delay - I changed the way I get Github notifications and I missed this. I added a couple of comments to the spreadsheet. I like where it's going, less Western-centric labels to name parts.  A couple of things popped out:

1.  Which field will single names be placed in? We should have a tool tip or something that helps our folks entering data that if you have a single name (e.g., [Nemayaq](https://arctos.database.museum/agents.cfm?agent_id=21280056)) that's in "given name" or "family name".
2.  will we still have aka fields for name variants? Pretty important field that could really benefit from the attribute metadata fields to explain when & how to use.
3.  I often use "shipping" address to indicate a place when I only know a community where someone is from (e.g., Nemayaq was from Camp Collier. That was a small Indigenous settlement on St. Lawrence Island and had no western-style addresses. It would be great to have a standardized option for indicating this sort of "address" for an individual in those situations.
4. We create a lot of non-person agents. Do we have a preferred way to enter those names? We've often had trouble with businesses who merge with other businesses or get bought out but retain some semblance of their name. It would be great if we could come up with some structure that clearly delineates these corporate genealogies.

That's all I can think of for now! Thanks - I look forward to playing around with some of these new attributes to see how we can more accurately describe people and organizations and their personal histories.

**Comment by dustymc** [comment]
> less Western-centric labels

Sorta, ``tên: Hồ Chí Minh`` is completely possible both in the current model and where we're going. (I'm not saying 'name' in every language as a label is a good idea - it's almost certainly not! - but it is and will remain possible.)

>Which field will single names be placed in?

Depends on the type, https://arctos.database.museum/info/ctDocumentation.cfm?table=ctagent_name_type will be my default mapping and definition.

>will we still have aka fields for name variants?

See above, my default will be a straight map. Feel free to suggest other stuff though!

>standardized option for indicating this sort of "address"

Ditto, my default will be a straight mapping using the terms and definitions from https://arctos.database.museum/info/ctDocumentation.cfm?table=ctaddress_type. (And somewhere I opened a rejected issue suggesting we do something different for those kinds of data...)

>Do we have a preferred way to enter those names? 

No change, but....

>merge with other businesses or get bought out but retain some semblance of their name.

....there's more metadata, including an agent relationship on every term - I don't know how precisely that might be handled, but 'better than now' seems straightforward.

One really huge (I think...) change will be less **_demand_** for more data from the system. (This is after all essentially a reaction to the rejected 'low-information to verbatim and  reserve Agent-Objects for things that need them' experiment.) If you want to create duplicates "A & B" and "A and B" and "A / B" and .... then you'll absolutely be able to, and you probably shouldn't expect to get the kinds of warnings you do now. You will probably be able to enter 'Nemayaq' as a birth date. (I could probably "require" some stuff, but I think I hear that's not wanted at the moment.) This model primarily changes the **_expectations_** of the system. (Which I suppose has nothing to do with attributes, that could happen in most any model, but changing to the more-flexible structure provides a handy opportunity to revisit the rules which have been put up with different expectations in mind.)



> playing around with some of these new attributes

Super-vague plan is to get what's in test to production ASAP (maybe tonight, if nobody has problems with the UI) and then start mapping things in test. Certainly some things aren't likely to become clear until I get a little deeper into that process, but I can definitely create attributes in test which you can play with if that might help clarify anything.

https://github.com/ArctosDB/arctos/issues/6813 and the roadmap linked from within it provide a bigger picture and the background leading here.

**Comment by Jegelewicz** [comment]
> Which field will single names be placed in? We should have a tool tip or something that helps our folks entering data that if you have a single name (e.g., [Nemayaq](https://arctos.database.museum/agents.cfm?agent_id=21280056)) that's in "given name" or "family name".

I don't think we should tell anyone what to do here. A single name could be either of these and we don't know that, the person entering data should. If it is not clear - then that name should just be "display name" with none of the other attributes supplied.

To be clear - once we move to the attribute model and remove the preferred name as key, I think one of our name types will need to be display - the string that will appear on Arctos webpages and in downloaded data. Required for all agents.

**Comment by dustymc** [comment]
>If it is not clear - then

... we have bad definitions - file an Issue!

> think one of our name types will need to be display

You're going to have to draw that out if you're proposing something. (My plan is to retain table ``agent`` with ``agent_id FKEY`` and ``preferred_agent_name varchar not null`` (that's an internal label, we can call it whatever in the UI, but I'm never going to be much fan of arbitrary changes...).

>Required

That's why ^^ - it's not really possible to require child data, surely we can agree that Agents need at least a string to exist, that's the model.

**Comment by dustymc** [comment]
>only know a community

general address type discussion: https://github.com/ArctosDB/arctos/issues/4560

**Comment by Jegelewicz** [comment]
> agent_id FKEY and preferred_agent_name varchar not null (that's an internal label, we can call it whatever in the UI, but I'm never going to be much fan of arbitrary changes...).

We can keep calling it preferred name, but I think that will be confusing once it isn't the key...but I could be wrong, and honestly I don't care as long as everyone can understand what it is/does.

**Comment by dustymc** [comment]
preferred_agent_name==>"call me John Doe"

is the key==>"Sorry, we've already got one, try again please" <-- we are here

isn't the key==>"OK" <-- we are going here
